asdf_compression/blsc/compressor.py

```diff
@@ -7,10 +7,34 @@ class BloscCompressor(Compressor):
     def compress(self, data, **kwargs):
         import blosc
 
-        yield blosc.compress(data, **kwargs)
+        # Type size, necessary for shuffle filters
+        typesize = data.itemsize
```

Comment:
Good catch! It appears that blosc [defaults to 8](https://www.blosc.org/python-blosc/reference.html#blosc.compress) whereas here (with the memoryview) the itemsize ([based off the blosc examples](https://www.blosc.org/python-blosc/tutorial.html#supporting-the-buffer-interface)) should be 1.

Reply:
No, the `itemsize` should be the type of the data ultimately stored. This is an optimization hint for the shuffle operators to expose more regularity. For example, when `float32` elements are stored, the `itemsize` should be 4, when `complex128` data are stored, it should be 16 etc.

I'm not familiar with all the details of `asdf` and using `data.itemsize` might be the wrong choice. I am looking for the datatype (that `asdf` knows) of the `ndarray` here.

Reply:
ASDF blocks are a little weird because they might be shared across arrays (even if the arrays have different dtypes). Because of this every block is just bytes.

Here's one example:
```python
import asdf, numpy as np
base_arr = np.arange(10, dtype='uint8')
view1 = base_arr.view(dtype='uint32')
view2 = base_arr.view(dtype='uint64')
asdf.AsdfFile({'v1': view1, 'v2': view2}).write_to('test.asdf')
```
The example produces a file with only 1 ASDF block but with 2 arrays each with a different dtype:
```yaml
#ASDF 1.0.0
#ASDF_STANDARD 1.5.0
%YAML 1.1
%TAG ! tag:stsci.edu:asdf/
--- !core/asdf-1.1.0
asdf_library: !core/software-1.0.0 {author: The ASDF Developers, homepage: 'http://github.com/asdf-format/asdf',
  name: asdf, version: 3.0.1.dev0}
history:
  extensions:
  - !core/extension_metadata-1.0.0
    extension_class: asdf.extension._manifest.ManifestExtension
    extension_uri: asdf://asdf-format.org/core/extensions/core-1.5.0
    software: !core/software-1.0.0 {name: asdf, version: 3.0.1.dev0}
v1: !core/ndarray-1.0.0
  source: 0
  datatype: uint32
  byteorder: little
  shape: [8]
v2: !core/ndarray-1.0.0
  source: 0
  datatype: uint64
  byteorder: little
  shape: [4]
...
```
See this issue for some discussion on asdf saving 'base' arrays: https://github.com/asdf-format/asdf/issues/1669

The standard is not very detailed about this behavior but does contain a short statement about [blocks being bytes and not having a datatype](https://asdf-standard.readthedocs.io/en/1.0.3/file_layout.html#blocks):

> Blocks represent a contiguous chunk of binary data and nothing more. Information about how to interpret the block, such as the data type or array shape, is stored entirely in ndarray structures in the tree, as described in [ndarray](https://asdf-standard.readthedocs.io/en/1.0.3/generated/stsci.edu/asdf/core/ndarray-1.0.0.html#core-ndarray-1-0-0).

Is there much of a performance gain to be had for setting `itemsize`?